### PR TITLE
DOCSP-30544: compound operations (crud)

### PR DESCRIPTION
**Pull request by rustagir**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-rust/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-30544>
**Staging**
- [page](https://docs-mongodbcom-staging.corp.mongodb.com/rust/docsworker-xlarge/DOCSP-30544-compound-crud/fundamentals/crud/compound-operations/) + see LH nav
- [crud landing](https://docs-mongodbcom-staging.corp.mongodb.com/rust/docsworker-xlarge/DOCSP-30544-compound-crud/fundamentals/crud/)

[build](https://workerpool-boxgs.mongodbstitch.com/pages/job.html?collName=queue&jobId=64c9733c87b3b360460a2cb9)

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/fundamentals/crud/compound-operations.txt [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+
+You can perform the following compound operations with the driver:
+
+- Find and delete one document
+- Find and update one document
+- Find and replace one document
+
+This guide includes the following sections:
+
+- :ref:`Sample Data for Examples <rust-compound-sample-data>`: presents
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:35]:
```suggestion
- :ref:`Sample Data for Examples <rust-compound-sample-data>`: Presents
```

applies to all items in this list

File: source/fundamentals/crud/compound-operations.txt [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+.. _rust-find-and-delete:
+
+Find and Delete a Document
+--------------------------
+
+The ``find_one_and_delete()`` method finds and deletes the first document that
+matches the specified query filter. If a document matches the filter
+criteria, the method returns a ``Some`` type. If it does not match any
+documents, it returns a ``None`` type.
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:83]: I: The document does the matching, not the method. Applies to whole page

S:

```suggestion
criteria, the method returns a ``Some`` type. If no documents match,
it returns a ``None`` type.
```

File: source/fundamentals/crud/compound-operations.txt [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+
+.. note:: 
+
+   If you don't need to find and delete a document in a single
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:87]: Q: Is there any advantage to calling them separately? Doing other processing between the method calls, maybe? Just don't want to confuse people about which way is preferred.

Applies to other notes on page too.

File: source/fundamentals/crud/compound-operations.txt [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+   operation, you can call the ``find_one()`` method followed by the
+   ``delete_one()`` method.
+
+Modify Find and Delete Behavior
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can optionally modify the behavior of the ``find_one_and_delete()``
+method by passing a ``FineOneAndDeleteOptions`` instance as a parameter.
+To use default values for each setting, specify the value ``None`` as
+the options parameter.
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:97]:
```suggestion
To use default values for each setting, specify the value ``None`` for
the options parameter.
```

Applies to same paragraph in each page section too

File: source/fundamentals/crud/compound-operations.txt [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+.. list-table::
+   :widths: 30 70
+   :header-rows: 1
+
+   * - Option
+     - Description
+
+   * - ``max_time`` 
+     - | The maximum amount of time, in milliseconds, that the query can
+         run for.
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:111]:
```suggestion
         run.
```

Applies in later paragraphs too

File: source/fundamentals/crud/compound-operations.txt [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+       | Type: ``Duration``
+
+   * - ``projection`` 
+     - | The projection to use when returning results.
+       | Default: ``nil``
+
+   * - ``sort`` 
+     - | The sort to use when returning results. By default, the driver
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:119]:
```suggestion
     - | The sorting order[?] to use when returning results. By default, the driver
```

Applies to whole page

File: source/fundamentals/crud/compound-operations.txt [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+
+You can optionally modify the behavior of the ``find_one_and_delete()``
+method by passing a ``FineOneAndDeleteOptions`` instance as a parameter.
+To use default values for each setting, specify the value ``None`` as
+the options parameter.
+
+The following table describes the options available in
+``FineOneAndDeleteOptions``:
+
+.. list-table::
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:102]: I: Some entries in this table are missing types; others are missing default values. Can you double-check?

Also applies to later tables on page

File: source/fundamentals/crud/compound-operations.txt [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+     - | The index to use for the operation. To learn more about
+         indexes, see :manual:`Indexes </indexes/>` in the Server
+         manual. This option is available only when connecting to
+         {+server+} versions 4.4 and later.
+       | Type: ``Hint``
+       | Default: ``nil``
+
+   * - ``let_vars``
+     - | A map of parameters and values. These parameters can be accessed
+         as variables in aggregation expressions. This option is available
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:146]: I: use active voice

S: Rephrase second sentence in active voice, with who/what is accessing the variables as the subject

Applies to same entry in later tables too

File: source/fundamentals/crud/compound-operations.txt [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+   * - ``comment``
+     - | An arbitrary ``Bson`` value tied to the operation to trace
+         it through the database profiler, ``currentOp``, and
+         logs. This option is available only when connecting to
+         {+server+} versions 4.4 and later.
+       | Type: ``Bson``
+
+.. TODO add links to guides for relevant options
+
+.. note:: Instantiating Options
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:159]: S: Delete this note, given the following example

Applies to later sections too

File: source/fundamentals/crud/compound-operations.txt [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+        "name": "Toby Fletcher",
+        "school": "Durango High School"
+      }
+
+.. _rust-crud-compound-addtl-info:
+
+Additional Information
+----------------------
+
+To learn more about performing the operations mentioned, see the
```
Comment (mongoKart) [inline source/fundamentals/crud/compound-operations.txt:526]:
```suggestion
To learn more about performing the previous operations, see the
```

**Review by mongoKart — CHANGES_REQUESTED** [review]
looking good. some small changes throughout

File: source/fundamentals/crud/compound-operations.txt (hunk 1 above)
Comment (rustagir) [reply source/fundamentals/crud/compound-operations.txt:35]: I think you and chris followed up about this rule, and it looked like items should be lowercased

File: source/fundamentals/crud/compound-operations.txt (hunk 2 above)
Comment (rustagir) [reply source/fundamentals/crud/compound-operations.txt:83]: changed in the three sections

File: source/fundamentals/crud/compound-operations.txt (hunk 3 above)
Comment (rustagir) [reply source/fundamentals/crud/compound-operations.txt:87]: yes, you risk data changes between calls by separating the calls

File: source/fundamentals/crud/compound-operations.txt (hunk 7 above)
Comment (rustagir) [reply source/fundamentals/crud/compound-operations.txt:102]: I havent been able to locate some of the default values, but i can definitely add types for all of the entries in every table.

File: source/fundamentals/crud/compound-operations.txt (hunk 9 above)
Comment (rustagir) [reply source/fundamentals/crud/compound-operations.txt:159]: Don't you think mentioning the design pattern might be useful to users who see the example and dont know whats going on?

Comment (mongoKart) [reply source/fundamentals/crud/compound-operations.txt:161]: Sure, that makes sense. What do you think about taking it out of an admonition and just letting it be text?

**Review by mongoKart — APPROVED** [review]
LGTM + 1 suggestion

File: source/includes/fundamentals/code-snippets/crud/compound.rs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+async fn main() -> mongodb::error::Result<()> {
+    let uri = "<connection string>";
+    let client = Client::with_uri_str(uri).await?;
+
+    let my_coll: Collection<Document> = client.database("db").collection("students");
+
+    let filter = doc! { "name": "xxx" };
+    // begin-find-delete-options
+    let opts = FindOneAndDeleteOptions::builder().comment(bson!("hello")).build();
+    let _res = my_coll.find_one_and_delete(filter, opts).await?;
```
Comment (kevinAlbs) [inline source/includes/fundamentals/code-snippets/crud/compound.rs:24]: Suggest removing `_` prefix from example snippet. I expect user may use `res` to check result. Same comment applies to other `_res` declarations.

File: source/includes/fundamentals/code-snippets/crud/compound.rs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    // end-find-delete-options
+
+    // begin-find_one_and_delete
+    let filter = doc! { "age": doc! { "$lte": 10 } };
+
+    let res = my_coll.find_one_and_delete(filter, None).await?;
+    println!("Deleted document:\n{}", serde_json::to_string_pretty(&res).unwrap());
```
Comment (kevinAlbs) [inline source/includes/fundamentals/code-snippets/crud/compound.rs:31]: `unwrap` panics on error. Suggest avoiding `unwrap` for consistent error handling.

Suggest printing the `Option<Document>` directly:

```
println!("Deleted document:\n{:?}", res);
```

Same comment applies to other calls of `unwrap`.

**Review by kevinAlbs — COMMENTED** [review]
LGTM with suggestions.

**Pull request merged**